### ISSUE TITLE
macOS BigSur Support

### ISSUE THREAD

**Issue by kognat-docs** [freemocap/freemocap#59]
Seems like the threading is the cause of the issue here.

```
(freemocap) MacBook-Pro:freemocap sam$ python runme_FreeMoCap.py 
Starting initialization for stage 1
  0%|                                                    | 0/20 [00:00<?, ?it/s]Oct  2 15:40:31  ThetaUVC_blender[33664] <Notice>: ------------ ThetaUVC_blender plugin start (version:2.0.1 built:Fri Aug 19 15:54:46 JST 2016 pid=33664 RELEASE). ------------ #thetauvc
  5%|██▏                                         | 1/20 [00:01<00:37,  1.98s/it]OpenCV: out device of bound (0-0): 1
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 2
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 3
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 4
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 5
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 6
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 7
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 8
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 9
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 10
OpenCV: camera failed to properly initialize!
 55%|███████████████████████▋                   | 11/20 [00:02<00:01,  7.18it/s]OpenCV: out device of bound (0-0): 11
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 12
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 13
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 14
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 15
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 16
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 17
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 18
OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0): 19
OpenCV: camera failed to properly initialize!
100%|███████████████████████████████████████████| 20/20 [00:02<00:00,  9.17it/s]
2021-10-02 15:40:41.571 python[33664:261625353] WARNING: NSWindow drag regions should only be invalidated on the Main Thread! This will throw an exception in the future. Called from (
	0   AppKit                              0x00007fff22b6ded1 -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 352
	1   AppKit                              0x00007fff22b58aa2 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1296
	2   AppKit                              0x00007fff22b5858b -[NSWindow initWithContentRect:styleMask:backing:defer:] + 42
	3   AppKit                              0x00007fff22e6283c -[NSWindow initWithContentRect:styleMask:backing:defer:screen:] + 52
	4   cv2.cpython-37m-darwin.so           0x000000010ead0c94 cvNamedWindow + 564
	5   cv2.cpython-37m-darwin.so           0x000000010eacdd1a _ZN2cv11namedWindowERKNS_6StringEi + 58
	6   cv2.cpython-37m-darwin.so           0x000000010dc4a487 _ZL23pyopencv_cv_namedWindowP7_objectS0_S0_ + 231
	7   python                              0x0000000104cd3f2b _PyMethodDef_RawFastCallKeywords + 395
	8   python                              0x0000000104e0b9bb call_function + 251
	9   python                              0x0000000104e034eb _PyEval_EvalFrameDefault + 20171
	10  python                              0x0000000104cd3b15 _PyFunction_FastCallKeywords + 229
	11  python                              0x0000000104e0b977 call_function + 183
	12  python                              0x0000000104e03462 _PyEval_EvalFrameDefault + 20034
	13  python                              0x0000000104cd3b15 _PyFunction_FastCallKeywords + 229
	14  python                              0x0000000104e0b977 call_function + 183
	15  python                              0x0000000104e03462 _PyEval_EvalFrameDefault + 20034
	16  python                              0x0000000104cd3b15 _PyFunction_FastCallKeywords + 229
	17  python                              0x0000000104e0b977 call_function + 183
	18  python                              0x0000000104e03462 _PyEval_EvalFrameDefault + 20034
	19  python                              0x0000000104cd1fea _PyFunction_FastCallDict + 234
	20  python                              0x0000000104cd66ba method_call + 122
	21  python                              0x0000000104cd445f PyObject_Call + 127
	22  python                              0x0000000104ef7e3a t_bootstrap + 122
	23  python                              0x0000000104e7c764 pythread_wrapper + 36
	24  libsystem_pthread.dylib             0x00007fff2031a8fc _pthread_start + 224
	25  libsystem_pthread.dylib             0x00007fff20316443 thread_start + 15
)
__________________________________________
cv2::videocapture properties for Camera# 0
CV_CAP_PROP_FRAME_WIDTH: '1280.0'
CV_CAP_PROP_FRAME_HEIGHT : '720.0'
CAP_PROP_FPS : '30.0'
CAP_PROP_EXPOSURE : '0.0'
CAP_PROP_POS_MSEC : '0.0'
CAP_PROP_FRAME_COUNT  : '0.0'
CAP_PROP_BRIGHTNESS : '0.0'
CAP_PROP_CONTRAST : '0.0'
CAP_PROP_SATURATION : '0.0'
CAP_PROP_HUE : '0.0'
CAP_PROP_GAIN  : '0.0'
CAP_PROP_CONVERT_RGB : '0.0'
__________________________________________
2021-10-02 15:40:42.898 python[33664:261625353] WARNING: nextEventMatchingMask should only be called from the Main Thread! This will throw an exception in the future.
```

**Comment by kognat-docs** [comment]
Seems this has some insight


https://github.com/kanishkaganguly/OpenNIMultiSensorCapture/blob/master/capture_cam.py

**Comment by kognat-docs** [comment]
This is useful

https://github.com/nrsyed/computer-vision/tree/master/multithread

also see:
https://nrsyed.com/2018/07/05/multithreading-with-opencv-python-to-improve-video-processing-performance/

**Comment by aaroncherian** [comment]
Interesting, mainly because I didn't write any threading into the code here, so I'm not sure where the threading is happening that would be causing an issue. At this point in the code it's running a for loop using OpenCV to try and open a camera at ports 0-20, checking which ones actually have a camera attached to them. 

Out of curiosity, did you have more than one camera plugged in? It seems like it found a webcam at port 0, and then didn't find one at any of the other ports.

 On a Windows system I've never seen that _"OpenCV: camera failed to properly initialize!
OpenCV: out device of bound (0-0):"_ statement pop up. I'm wondering if it might be a Mac-specific error message.

**Comment by kognat-docs** [comment]
The following `.diff` fixes it

```
diff --git a/freemocap/runcams.py b/freemocap/runcams.py
index 5446d25..b153f2c 100644
--- a/freemocap/runcams.py
+++ b/freemocap/runcams.py
@@ -42,13 +42,13 @@ def RecordCams(session,camInputs,parameterDictionary,rotationInputs):
             beginTime,
             parameterDictionary,
         )
-        camRecordings.start()
+        camRecordings.run()
 
         threads.append(camRecordings)
 
     for camRecordings in threads:
-        camRecordings.join()  # make sure that one thread ending doesn't immediately end all the others (before they can dump data in a pickle file)
-
+        #camRecordings.join()  # make sure that one thread ending doesn't immediately end all the others (before they can dump data in a pickle file)
+        pass
     print("finished recordings")
 
     timeStampList = []
diff --git a/freemocap/webcam/camsetup.py b/freemocap/webcam/camsetup.py
index b27b096..32df333 100644
--- a/freemocap/webcam/camsetup.py
+++ b/freemocap/webcam/camsetup.py
@@ -4,7 +4,7 @@ import imutils
 import os
 
 
-class VideoSetup(threading.Thread):
+class VideoSetup():
     """
     Class to run and thread webcams for preview purposes
     """
@@ -12,7 +12,7 @@ class VideoSetup(threading.Thread):
         self.camID = camID
         self.parameterDictionary = parameterDictionary
         self.rotNum = rotNum
-        threading.Thread.__init__(self)
+        #threading.Thread.__init__(self)
 
     def run(self):
         # print("Starting " + self.previewName)
@@ -78,8 +78,9 @@ def RunSetup(cam_inputs, rotation_input, paramDict):
 
     for cam_input, cam_rotation in zip(cam_inputs, rotation_input):
         u = VideoSetup(cam_input, paramDict, cam_rotation)
-        u.start()
+        u.run()
         ulist.append(u)
 
     for k in ulist:
-        k.join()
+        #k.join()
+        pass
diff --git a/freemocap/webcam/startcamrecording.py b/freemocap/webcam/startcamrecording.py
index 3a791cc..45caf5d 100644
--- a/freemocap/webcam/startcamrecording.py
+++ b/freemocap/webcam/startcamrecording.py
@@ -4,11 +4,11 @@ import time
 import pickle
 import os
 
-class CamRecordingThread(threading.Thread):
+class CamRecordingThread():
     def __init__(
         self, session, camID, camInput, videoName, rawVidPath, beginTime, parameterDictionary
     ):
-        threading.Thread.__init__(self)
+        #threading.Thread.__init__(self)
         self.camID = camID
         self.camInput = camInput
         self.videoName = videoName

```

**Comment by kognat-docs** [comment]
I will try it with two extra cameras plugged in now I will need to borrow a laptop to use a charuco board.

**Comment by kognat-docs** [comment]
Works OK 

my cameras are difficult to setup well using a Charuco board on my phone

But no blocking so far

**Comment by kognat-docs** [comment]
Actually works fine for setup

But records fails

If it is all on the main thread you need to have an event loop also on that main thread

I will see if I get some time later this week or next week

**Comment by kognat-docs** [comment]
Will try asyncio to see if I can have a bit of paralell action.

**Comment by kognat-docs** [comment]
This might help convert the original program

https://stackoverflow.com/questions/57967745/convert-a-simple-multithreaded-program-with-asyncio

**Comment by kognat-docs** [comment]
Might need to do a simple asyncio program that captures both the .csv file of time stamps and the frames.

Then look at interfaces to connect with the rest framework so Linux and MacOS can work without the existing issues with threading

**Comment by FernandaBernardo** [comment]
I have the same problem but using macOS Monterey M1. But I think the project has changed a lot since this issue was opened. Can anybody help me?

**Comment by skylarpape** [comment]
Having the same issue. @jonmatthis is support for macOS something the project would consider prioritizing or accepting contributions for?

**Comment by trentwirth** [comment]
TLDR: freemocap should run on macOS, Ubuntu, and Windows soon.

Our alpha code (which we're currently testing for release) should run on macOS, Ubuntu, and Windows. 

We'll be having a bigger release of alpha soon. It takes a lot of work, but you can play around with our Alpha now if you're tech savvy (Jon does a fast-paced demo of it in this video: https://www.youtube.com/watch?v=0k-jvAz8NDI ) 

If you're not tech savvy, when our alpha is released it will be at least as easy to launch as the pre-alpha, hopefully easier (as a downloadable `.exe` from our github)
